Docs/man/gnokiid.8
.TH "gnokiid" "8" "May 27, 2001" "Erik Rossen" "gnokii"
.SH "NAME"
gnokiid \- create a virtual modem for old Nokia mobile phones
.SH "SYNOPSIS"
.B gnokiid

.B gnokiid
[\fB\-\-help\fR|\fB\-\-version\fR|\fB\-\-debug\fR]

.SH "DESCRIPTION"
.PP 
.B gnokiid

is a tool from the gnokii tool suite used to create a virtual modem device
with a Nokia portable telephone.  The virtual modem device is usually
accessed via a symbolic link called /dev/gnokii (created automatically by
\fBmgnokiidev\fR(8)). Please note that you don't need a virtual modem when
you have real modem accepting AT Hayes commands (most of the modern
phones). gnokiid was written for the phones that supported data transfer
but didn't expose AT layer (namely Nokia 6110, 6130, 6150, 5110, 5130).

.PP 
After 
.B gnokiid
has created the virtual modem, it can be accessed like any other Hayes\-compatible modem by such programs as the terminal program \fBminicom\fR(1) or the PPP daemon \fBpppd\fR(8) to make standard data connections to remote computers.

.PP 
The full range of portable telephone functions (SMS messaging, phone book, operator logos, etc.) are accessed using the command\-line program \fBgnokii\fR(1) or the graphical interface program \fBxgnokii\fR(1x).


.PP 
gnokii
is a Linux/Unix tool suite and (eventually) modem/fax driver for the mobile phones, released under the GPL.
.PP 
gnokii
supports most phones from the 3810/8110 and 5110/6110 series, details including bugs specific to each series appear in the files Docs/README\-3810 and Docs/README\-6110 respectively.  

.SH "OPTIONS"
.PP 
If called without any options, \fBgnokiid\fR will set up a virtual modem device using the settings in gnokii(1) config file (if it exists).  Otherwise,
.TP 
\fB\-\-help\fR
print out a short usage message
.TP 
\fB\-\-version\fR
print out version, copyright, phone, and serial device info
.TP 
\fB\-\-debug\fR
uses STDIN/STDOUT for a virtual modem
.SH "DIAGNOSTICS"
Various error messages are printed to STDERR.
.SH "BUGS"
.PP 
Hmmm...


.SH "AUTHOR"
Hugh Blemings <hugh at blemings dot org> and Pavel Janik ml. <Pavel.Janik at suse dot cz> are the authors of the
gnokii tool suite.

This manual page was written by Erik Rossen <rossen at freesurf dot ch>.

See also Docs/CREDITS from the gnokii sources.
.SH "COPYING"
This program is distributed under the GNU General Public License Version 2, or (at your option) any later version.

.SH "SEE ALSO"
gnokii, xgnokii, mgnokiidev


ChangeLog
0.6.32 (not released yet)
=========================
 * translation updates
    o updated Italian translation                   (Daniele Forsi)
    o updated French translation                     (Raphaël Droz)
 * libgnokii updates
    o replace gethostbyname (deprecated) with getaddrinfo in tcp
      driver                                            (Paweł Kot)
    o add optional authentication during the initialization phase
                                                        (Paweł Kot)
    o move device_script() to generic device; it will cause ability
      to execute connect_script and disconnect_script with all
      connection types                                  (Paweł Kot)
    o implement GN_OP_Ping operation                    (Paweł Kot)
    o move callbacks from gn_data to separate structure put into
      gn_statemachine                                   (Paweł Kot)
    o add id for Nokia 3720c to misc.c only; use model=series40
                                                     (Anders Jenbo)
    o add another id for Nokia 1110 and ids for Nokia 1112 and 1208
      to misc.c only; use model=series40; many commands don't work,
      see the Series 30 section in Docs/README-6510
                                                   (Gustavo Conrad)
    o remove inline from the local_atoi() function definition and
      add static instead                                (Paweł Kot)
    o don't overwrite the validity value passed by the caller to
      sms_prepare() with the default value          (Daniele Forsi)
 * at driver updates
    o in the default case autodetect if PDU SMS starts with SMSC
                                                    (Daniele Forsi)
    o fix parsing responses contining non-ascii characters
                                                        (Paweł Kot)
    o fix sending sms with Nokia Cellular Data Adapter; thanks to
      Tristan Drinkwater for patient testing            (Paweł Kot)
    o add optional authentication during the initialization phase
                                                        (Paweł Kot)
    o implement GN_OP_Ping operation                    (Paweł Kot)
 * fake driver updates
    o always autodetect if PDU SMS starts with SMSC (Daniele Forsi)
    o make it usable with gnome-phone-manager       (Daniele Forsi)
 * documentation updates
    o samples for connect and disconnect scripts for Portech
      devices                                   (Guillermo Benitez)
    o update man page with allowed values for --validity option of
      --sendsms                                     (Daniele Forsi)
 * nk6510 driver updates
    o store also the seconds in date fields when reading phonebook
      entries from phone                            (Daniele Forsi)
    o add support for phonebook memories SD (Service Dialling
      Numbers), MR (Message Recipients), CL (Call Log) and
      CB (Currently selected memory)                (Daniele Forsi)
    o parse call duration subblock found in Dialled Calls (DC),
      Received Calls (RC) and Call Log (CL) phonebook entries;
      output format is subject to change            (Daniele Forsi)
    o change the value used for phonebook memory ON (reading and
      deleting work, but writing fails)             (Daniele Forsi)
    o change the value used for VOICE phonebook memory, containing
      the phone number(s) to be used to listen to voicemail, which
      was used for MT which now is unsupported      (Daniele Forsi)
    o use libusb to detach cdc_phonet or any other linux kernel
      module; modprobe blacklisting isn't needed anymore
                                                    (Daniele Forsi)
 * xgnokii updates
    o fix bug that caused to forget subentries after the first one
      when editing the phonebook                    (Daniele Forsi)
    o change relevant inline defintion to static inline to follow
      C99 standard                                      (Paweł Kot)
 * nk7110 driver updates
    o return more specific error codes for GN_OP_ReadPhonebook and
      for GN_OP_GetSMSCenter failures               (Daniele Forsi)
    o add support for phonebook memory SD (Service Dialling
      Numbers) and CB (Currently selected memory)   (Daniele Forsi)
    o change the value used for VOICE phonebook memory, containing
      the phone number(s) to be used to listen to voicemail, which
      was used for MT which now is unsupported      (Daniele Forsi)

0.6.31
======
 * libgnokii updates
    o add gn_lib_phone_change_driver() function to allow replacing
      used driver in runtime                            (Paweł Kot)
    o GSM number type is rather mask than define; use it as such
      when parsing the input                            (Paweł Kot)
    o avoid overruns in phonet                          (Paweł Kot)
    o print names of config files it tried to open only if none
      could be used                                 (Daniele Forsi)
    o fix setting alarms for zero length events read from icalendar
      files (eg. birthdays)                         (Daniele Forsi)
    o fix sms encoding: don't break multibyte characters when
      creating multipart messages                       (Paweł Kot)
    o add possibility to disable setting serial port speed and
      DTR and RTS bits                                  (Paweł Kot)
    o change the order of lookup for the config file
      (http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=623694)
                                                        (Paweł Kot)
    o add id for Nokia 3109c and for Nokia C1-01 to misc.c only;
      both use model=series40     (Alexandre A. Alves, Josh Kaplan)
    o fix gn_sms_send() so the message could be resent  (Paweł Kot)
    o make the vcard reader handle also BDAY, TITLE, NICKNAME, ORG,
      X-SIP;POC and X-WV-ID                         (Daniele Forsi)
    o make the vcard reader handle also empty fields in N and in

gnokiid/gnokiid.c
/*

  G N O K I I

  A Linux/Unix toolset and driver for the mobile phones.

  This file is part of gnokii.

  Copyright (C) 1999-2000  Hugh Blemings & Pavel Janik ml.
  Copyright (C) 2001-2004  Pawel Kot
  Copyright (C) 2002       Manfred Jonsson
  Copyright (C) 2002-2003  BORBELY Zoltan

  Mainline code for gnokiid daemon. Handles command line parsing and
  various daemon functions.

*/

#include "config.h"

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <termios.h>
#include <fcntl.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/time.h>
#include <string.h>

#include "config.h"

#ifdef ENABLE_NLS
#  include <locale.h>
#endif

#include "misc.h"
#include "gnokii-internal.h"

/* Global variables */
bool DebugMode;		/* When true, run in debug mode */
const char *BinDir;	/* Directory of the mgnokiidev command */
char *lockfile = NULL;
bool GTerminateThread;


static void short_version()
{
	fprintf(stderr, _("GNOKIID Version %s\n"), VERSION);
}

static void version()
{
	fprintf(stderr, _("Copyright (C) Hugh Blemings <hugh@blemings.org>, 1999\n"
			  "Copyright (C) Pavel Janik ml. <Pavel.Janik@suse.cz>, 1999\n"
			  "Built %s %s\n"),
			  __TIME__, __DATE__);
}

/* The function usage is only informative - it prints this program's usage and
   command-line options.*/
static void usage()
{

	fprintf(stderr, _("   usage: gnokiid [--help|--version|--debug]\n"
"          --help            display usage information\n"
"          --version         displays version and copyright information\n"
"          --debug           uses stdin/stdout for virtual modem comms\n"));
}

/* cleanup function registered by atexit() and called at exit() */
static void busterminate(void)
{
	gn_vm_terminate();
	gn_device_unlock(lockfile);
	gn_lib_library_free();
}

/* Main function - handles command line arguments, passes them to separate
   functions accordingly. */

int main(int argc, char *argv[])
{
	const char *aux;
	static bool atexit_registered = false;
	struct gn_statemachine temp_state;

	/* For GNU gettext */
#ifdef ENABLE_NLS
	setlocale(LC_ALL, "");
	bindtextdomain(GETTEXT_PACKAGE, LOCALEDIR);
	textdomain(GETTEXT_PACKAGE);
#endif

	short_version();

	/* Handle command line arguments. */
	switch (argc) {
	case 1:
		/* Default */
		DebugMode = false;
		break;
	case 2:
		if (strcmp(argv[1], "--version") == 0) {
			/* Display version, copyright and build information. */
			version();
			exit(0);
		} else if (strcmp(argv[1], "--debug") == 0) {
			/* Use stdin/stdout for I/O */
			DebugMode = true;
			break;
		} else if (strcmp(argv[1], "--help") == 0) {
			usage();
			exit(0);
		}
		/* FALL TROUGH */
	default:
		usage();
		exit(1);
	}

	aux = gn_lib_cfg_get("global", "model");
	if (!aux) {
		fprintf(stderr, "%s\n", gn_error_print(GN_ERR_NOCONFIG));
		exit(1);
		}

	if (strncmp(aux, "5110", 4) &&
	    strncmp(aux, "5130", 4) &&
	    strncmp(aux, "6110", 4) &&
	    strncmp(aux, "6130", 4) &&
	    strncmp(aux, "6150", 4)) {
		fprintf(stderr, _("gnokiid purpose is to work only with the phones that do not have AT Hayes\ncommands interpreter.\n"));
		exit(1);
	}

	BinDir = gn_lib_cfg_get("global", "bindir");
	if (!BinDir) BinDir = gn_lib_cfg_get("gnokiid", "bindir");
	if (!BinDir) BinDir = SBINDIR;

	if (gn_cfg_phone_load("", &temp_state) != GN_ERR_NONE) exit(-1);

	gn_elog_handler = NULL;

	if (temp_state.config.use_locking) {
		lockfile = gn_device_lock(temp_state.config.port_device);
		if (lockfile == NULL) {
			fprintf(stderr, _("Lock file error. Exiting.\n"));
			exit(1);
		}
	}

	/* register cleanup function */
	if (!atexit_registered) {
		atexit_registered = true;
		atexit(busterminate);
	}

	while (1) {
		if (gn_vm_initialise("", BinDir, DebugMode, true) == false) {
			exit (-1);
		}

		GTerminateThread = false;

		gn_vm_loop();
	}

	exit (0);
}
